// as-proto-gen/lib/scope-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeContext = void 0;
const reserved_keywords_1 = require("./reserved-keywords");
class ScopeContext {
    constructor(fileContext, reservedNames = []) {
        this.fileContext = fileContext;
        this.registeredNames = new Set(reservedNames);
    }
    /**
     * Return file context in which the scope is.
     */
    getFileContext() {
        return this.fileContext;
    }
    /**
     * Registers a new name in a scope.
     */
    registerName(preferredName) {
        let freeName = (0, reserved_keywords_1.getSafeName)(preferredName);
        let freeSuffix = 2;
        while (this.registeredNames.has(freeName)) {
            freeName = `${preferredName}_${freeSuffix++}`;
        }
        this.registeredNames.add(freeName);
        return freeName;
    }
    /**
     * Checks if given name is registered in a scope.
     */
    hasRegisteredName(name) {
        return this.registeredNames.has(name);
    }
}
exports.ScopeContext = ScopeContext;
